src/main/scala/inox/ast/Paths.scala

```diff
@@ -185,27 +181,37 @@ trait Paths { self: SymbolOps with TypeOps =>
       */
     def filterByIds(ids: Set[Identifier]): Path = {
       def containsIds(ids: Set[Identifier])(e: Expr): Boolean = exprOps.exists {
-        case Variable(id, _, _) => ids.contains(id)
+        case Variable(id, _, _) => ids contains id
         case _ => false
       }(e)
 
-      val newElements = elements.filter{
-        case Left((vd, e)) => ids.contains(vd.id) || containsIds(ids)(e)
-        case Right(e) => containsIds(ids)(e)
+      val newElements = elements filter {
+        case CloseBound(vd, e) => (ids contains vd.id) || containsIds(ids)(e)
+        case OpenBound(vd) => ids contains vd.id
+        case Condition(e) => containsIds(ids)(e)
       }
-      new Path(newElements, bounds)
+
+      new Path(newElements)
     }
 
     /** Free variables within the path */
-    lazy val variables: Set[Variable] = fold[Set[Variable]](Set.empty,
-      (vd, e, res) => res - vd.toVariable ++ exprOps.variablesOf(e), (e, res) => res ++ exprOps.variablesOf(e)
-    )(elements)
+    lazy val freeVariables: Set[Variable] = {
```

Comment:
I like the name `variables`, it fits with `variablesOf`, but I understand if you want to keep `freeVariables`.

Reply:
I've kept `freeVariables` because it's confusing otherwise. ;-)